### PR TITLE
Configuration: change load order (system properties and environment variables first)

### PR DESCRIPTION
**Pull request by gonzalad**
Hello,

**Summary**: This PR will change the load order of CAS configuration so that System properties and env variables
override CAS configuration files.

**Details**:

When I'm using CAS in a docker environment (with a docker-compose file), I stumbled
upon the fact that I could not override a variable already defined in /etc/cas/config/application.yml
from my docker-compose file (i.e I'm used to override the spring boot configuration with environment variables defined in the docker compose)

For instance, if I have:

{code}
version: '3'
services:
  cas:
    image: apereo/cas
    ports:
      - "8443:8443"
      - "8444:8444"
    volumes:
      - ./etc/cas/config:/etc/cas/config
    environment:
      - SERVER_PORT=8444
{code}

With my /etc/cas/config/application.yml being

{code}
server:
  port: 8443
{code}

The CAS will start at port 8443 (and not 8444 as I would expect from a Spring Boot app)

This PR will change the load order of CAS configuration so that System properties and env variables
override CAS configuration files.

N.B. I've added a unit test for checking system.properties configuration order, but I didn't found a good way to test environment variables ordering. I'll take any input gladly.

### CODE REVIEW

File: core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java [hunk 1]
```diff
@@ -168,6 +175,23 @@
         return composite;
     }
 
+    /**
+     * Returns a system property composed of environment variables and system properties.
```
Comment (mmoayyed) [inline core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java:179]: You're not returning a system property; You're returning a property source.

File: core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java [hunk 2]
```diff
@@ -168,6 +175,23 @@
...
+     *
+     * <p>Properties present in system properties will
+     * take precedence over those in environment variables,
+     * similarly to spring boot behaviour.
```
Comment (mmoayyed) [inline core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java:183]: Please switch to `behavior`. Also remove the initial `<p>` tag, if possible.

File: core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java [hunk 3]
```diff
@@ -11,16 +11,21 @@
...
 import org.springframework.core.io.Resource;
 import org.springframework.core.io.ResourceLoader;
 
 import java.io.File;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.stream.Collectors;
+import static org.springframework.core.env.StandardEnvironment.SYSTEM_ENVIRONMENT_PROPERTY_SOURCE_NAME;
```
Comment (mmoayyed) [inline core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java:27]: Please remove static imports, and instead use `StandardEnvironment.SYSTEM_ENVIRONMENT_PROPERTY_SOURCE_NAME` and the like.

File: core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java [hunk 4]
```diff
@@ -168,6 +175,23 @@
...
+    /**
+     * Returns a system property composed of environment variables and system properties.
+     *
+     * <p>Properties present in system properties will
+     * take precedence over those in environment variables,
+     * similarly to spring boot behaviour.
+     */
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    private PropertySource<?> loadEnvironmentAndSystemProperties() {
+        CompositePropertySource environmentAndSystemProperties = new CompositePropertySource("environmentAndSystemProperties");
```
Comment (mmoayyed) [inline core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java:187]: Pleas run the style guide to make sure you're up to speed.

**Review by mmoayyed — CHANGES_REQUESTED** [review]

File: core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java (hunk 1 above)
Comment (gonzalad) [reply core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java]: I just changed the javadoc to:

```
     * Returns a property source composed of system properties and environment variables.
     *
     * System properties take precedence over environment variables (similarly to spring boot behaviour).
```

Let me know if it's ok to you

File: core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java (hunk 2 above)
Comment (gonzalad) [reply core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java]: See previous comment (I didn't understood what exactly you mean by behaviour - so I hope it's ok ;) )

File: core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java (hunk 4 above)
Comment (gonzalad) [reply core/cas-server-core-configuration-api/src/main/java/org/apereo/cas/configuration/DefaultCasConfigurationPropertiesSourceLocator.java]: Done, I added the static modifier

**Comment by gonzalad** [comment]
@mmoayyed I made the changed you asked, but I don't know if it's up to me to click on the `Resolve conversation` buttons

**Comment by mmoayyed** [comment]
Thanks. Syncing is always Ok. I will take a look and should be able to merge quickly.

**Pull request merged**